src/main/java/vct/col/rewrite/ParallelBlockEncoder.java

```diff
@@ -922,7 +926,7 @@ protected void check_send_recv(DeclarationStatement[] body_decl,
         cb=new ContractBuilder();
         cb.requires(loop_invariant);
         cb.ensures(loop_invariant);
-        cb.requires(guard);
+        cb.requires(create.fold(StandardOperator.And, parBoundsStack));
```

Comment:
I'm a bit confused about what parBoundsStack does, it seems to collect all conditions of all foreachloops encountered on the way (so the "i < 10" part of the foreach loop, I imagine)? I guess the ParallelBlockEncoder is an enigma anyway, but if you can give me some insight here I'd be interested.

Reply:
ParallelBlockEncoder is to some degree aware of nested parallel blocks, propagating invariants and such correctly. Apparently to this day the bounds of parallel blocks were not considered invariants, so if you have something like:

```
par(int a = 0..N) {
  par(int b = 0..M) {}
}
```

the proof of the inner par may use `a` and is passed the value of `a`, but it loses the knowledge that `0 <= a < N`. This was the stumbling block for some proof I encountered and it seems obvious to include it...